[AArch64] Don't try to vectorize fixed point to fp narrowing conversion

pranavk opened the pull request:
GCC, correctly, doesn't vectorize in this case. Absence of direct instructions to convert larger fixed point to lower floating point precision inadvertently causes rounding leading to subtle differences across ISAs.

https://godbolt.org/z/ssEchMWrE

Co-authored by: @echristo

rnk commented:
Test case? Maybe you have one but forgot to `git add` the new file.

davemgreen commented:
There will be a lot of test cases that need updating. We should attempt to limit the damage this does, and I don't believe the default expansion is very efficient for unsigned operations.

echristo commented:
> There will be a lot of test cases that need updating. We should attempt to limit the damage this does, and I don't believe the default expansion is very efficient for unsigned operations.

I did notice that arguably turning off vectorization of this operation might be the most efficient way. It does leave the loop rather than unrolling it, but the code is quite a bit tighter and the scalar conversion is precise.

davemgreen commented:
> I did notice that arguably turning off vectorization of this operation might be the most efficient way. It does leave the loop rather than unrolling it, but the code is quite a bit tighter and the scalar conversion is precise.

Yeah - we should probably try to address both. There are two issues going on - no matter where the vector i64->f32 convert comes from (vectorization, intrinsics, optimizations, the front end) we should be producing correctly rounded code (*). This patch is the bug fix for that issue. Then there is the question of whether the vectorizer should be producing the vector operations. That will probably require a cost-model change to increase the costs (which might mean they are still used in loops with other vector operations, they will just be scalarized for that portion, it still might be profitable overall).

(* I was wondering if there was something that we could do with fast-math to make this OK. Especially as AFAICT it has been this way for 10+ years without anyone having an issue so far. None of the existing flags match exactly, but maybe contract or afn could be co-opted into it. (If fptrunc+fptrunc could be contracted, maybe fptrunc can be "uncontracted" into fptrunc+fptrunc). Something for a follow-on if we need it maybe, we might be able to use SVE instructions instead for some of them if the architecture feature is present).

If you can update all the failing test cases we can take a look at whether they need to change. I believe, for example, that i32->half conversions are always fine (https://alive2.llvm.org/ce/z/RXvYNv). That probably applies to higher int sizes too as the range or a half is relatively narrow. BFloats I would guess the same as float, unfortunately, but it would be good if we could prove one was right or wrong.

pranavk commented:
Thanks for taking a look.

I think you are right about ->half conversions. For now, I have made this conditional on target type being f32 to try to limit the impact, and also to avoid the backend crash for one of the tests (`llvm/test/CodeGen/AArch64/sitofp-to-tbl.ll:sitofp_v8i8_to_v8f16`). 

The diffs look pretty inefficient as we guessed.

For the test that crashes, backend crashes because sitofp's operand in this test (`v8i8`) gets promoted to `v8i32` instead of custom-lowered. So the place where I added `SDValue` is never encountered which in other cases leads to `expansion` of vector ops to scalar nodes. This leads of `v8i32`in the DAG by the time we call `SelectionDAG::Legalize()` which now encounters the `return SDValue` I introduced here which leads to trying to find an appropriate libcall for v4i32 -> v4f16 conversion which fails:

```
Assertion `LC != RTLIB::UNKNOWN_LIBCALL && "Unable to legalize as libcall"' failed
...
#10 0x000055b3966f8f03 (anonymous namespace)::SelectionDAGLegalize::ConvertNodeToLibcall(llvm::SDNode*) /usr/local/google/home/prka/p/llvm/llvm_main2/llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:4800:21
#11 0x000055b3966e7e92 (anonymous namespace)::SelectionDAGLegalize::LegalizeOp(llvm::SDNode*) /usr/local/google/home/prka/p/llvm/llvm_main2/llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:1353:7
#12 0x000055b3966e5f04 llvm::SelectionDAG::Legalize() /usr/local/google/home/prka/p/llvm/llvm_main2/llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:5861:13
#13 0x000055b396891413 llvm::SelectionDAGISel::CodeGenAndEmitDAG() /usr/local/google/home/prka/p/llvm/llvm_main2/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:1078:3
#14 0x000055b396890193 llvm::SelectionDAGISel::SelectBasicBlock(llvm::ilist_iterator_w_bits<llvm::ilist_detail::node_options<llvm::Instruction, true, false, void, true, llvm::BasicBlock>, false, true>, llvm::ilist_iterator_w_bits<llvm::ilist_detail::node_options<llvm::Instruction, true, false, void, true, llvm::BasicBlock>, false, true>, bool&) /usr/local/google/home/prka/p/llvm/llvm_main2/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:877:1
#15 0x000055b39688fd76 llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) /usr/local/google/home/prka/p/llvm/llvm_main2/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:1903:11
#16 0x000055b39688cd99 llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) /usr/local/google/home/prka/p/llvm/llvm_main2/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:614:3
#17 0x000055b39482de05 (anonymous namespace)::AArch64DAGToDAGISel::runOnMachineFunction(llvm::Mach
```

davemgreen commented on llvm/test/CodeGen/AArch64/complex-int-to-fp.ll:
```diff
@@ -1,9 +1,20 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
 ; RUN: llc < %s -mtriple=arm64-eabi -aarch64-neon-syntax=apple | FileCheck %s
 
 ; CHECK: autogen_SD19655
 ; CHECK: scvtf
 ; CHECK: ret
```
Remove these old check lines

davemgreen commented on llvm/test/CodeGen/AArch64/fp16-v16-instructions.ll:
```diff
@@ -23,26 +23,54 @@ define <16 x half> @sitofp_i32(<16 x i32> %a) #0 {
-; CHECK-NEXT:    scvtf v1.2d, v7.2d
-; CHECK-NEXT:    fcvtn2 v4.4s, v5.2d
-; CHECK-NEXT:    fcvtn2 v2.4s, v3.2d
+; CHECK-NEXT:    fmov x12, d0
```
Is it possible to have this still work as before? It is converting to a fp16, but probably doing so in multiple steps so there is still a i64->fp32 convert.

pranavk replied:
I tried looking. If we want to do that, the check likely needs to be moved earlier before LowerVectorINT_TO_FP (and I am not sure there's any appropriate position for that earlier) as there doesn't seem to be any way to distinguish between instructions that are generated by the compiler (during split, promote, etc.) vs original sint_to_fp instruction once we reach LowerVectorINT_TO_FP.

pranavk replied:
Okay, I pattern-matched that and was able to exclude all such tests that eventually end up rounding to f16 as that doesn't exhibit rounding errors we are worried about. That did get rid of ~5 tests that remain unchanged with this PR now.

pranavk replied:
I will refactor the existing code in a while. Perhaps there are elegant ways to pattern match a thing like this.

davemgreen reviewed (commented):
The unsigned variants are trying to do some sort of math to keep things in vectors, but it would be better to just scalarize as we have a scalar instruction and the number of vector lanes is only 2. Can we try and get it to scalarize instead?

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -5106,6 +5106,30 @@ SDValue AArch64TargetLowering::LowerVectorINT_TO_FP(SDValue Op,
   if (VTSize < InVTSize) {
+    // AArch64 doesn't have a direct vector instruction to convert
+    // fixed point to floating point AND narrow it at the same time.
+    // Additional rounding when the target is f32/f64 causes subtle
+    // differences across different platforms (that do have such
```
Perhaps "causes subtle differences across different platforms(that do have such instructions)" -> "causes double-rounding issues".

pranavk replied:
done

davemgreen reviewed (commented):
I've put together a patch in #132366 to update the costs. I will look into updating the GISel codegen too, if you are not looking into it. I am hoping it is just an update to the legalization descriptors for the operations.

pranavk committed:
init

pranavk committed:
modify tests

pranavk commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -5106,6 +5106,30 @@ SDValue AArch64TargetLowering::LowerVectorINT_TO_FP(SDValue Op,
+    }
+
+    if (IsTargetf32Orf64 && !IsTargetf16) {
+      return SDValue();
```
`UnrollVectorOp` doesn't seem to handle strict-fp opcodes gracefully. LLVM crashes if I do that. So I am still doing `SDValue()` for strict-fp opcodes.

davemgreen reviewed (approved):
I updated the complex-int-to-fp.ll test to remove the older CHECK lines, to make sure they shouldn't cause a problem.

LGTM, so long as the tests pass. Thanks

pranavk commented:
Rebased. If tests pass, I will merge this. 

Thank you for reviewing and other patches (cost-model, and GISel changes).
